estargz/build.go

```diff
@@ -267,26 +273,37 @@ func Build(tarBlob *io.SectionReader, opt ...Option) (_ *Blob, rErr error) {
 	}
 	diffID := digest.Canonical.Digester()
 	pr, pw := io.Pipe()
+	uncompressedSizeChan := make(chan int64, 1)
 	go func() {
-		r, err := opts.compression.Reader(io.TeeReader(io.MultiReader(append(rs, tocAndFooter)...), pw))
+		var uncompressedSize int64
+		var decompressFunc func(io.Reader) (io.ReadCloser, error)
+		if _, ok := opts.compression.(*gzipCompression); ok && opts.gzipHelperFunc != nil {
+			decompressFunc = opts.gzipHelperFunc
+		} else {
+			decompressFunc = opts.compression.Reader
+		}
+		decompressR, err := decompressFunc(io.TeeReader(io.MultiReader(append(rs, tocAndFooter)...), pw))
 		if err != nil {
 			pw.CloseWithError(err)
 			return
 		}
-		defer r.Close()
-		if _, err := io.Copy(diffID.Hash(), r); err != nil {
+		defer decompressR.Close()
+		if uncompressedSize, err = io.Copy(diffID.Hash(), decompressR); err != nil {
 			pw.CloseWithError(err)
 			return
 		}
+		uncompressedSizeChan <- uncompressedSize
+		close(uncompressedSizeChan)
```

Comment:
Does this make `UncompressedSize` return a zero value when it called more than once? And I guess this coroutine leaks with blocking at L295 if `UncompressedSize` isn't called. 
Instead of using channel, I think it can just use an int64 pointer + atomic. `UncompressedSize` can have a comment that the value is valid only after the full read.